ocaml/xapi/xapi_vdi.ml

```diff
@@ -878,26 +878,32 @@ let get_nbd_info ~__context ~self =
     raise (Api_errors.Server_error (Api_errors.vdi_incompatible_type, [ Ref.string_of self; Record_util.vdi_type_to_string `cbt_metadata ]))
   end;
 
+  (* Ignore exceptions that can happen due to invalid references and just skip them *)
+  let default_on_exn f default x = try f x with _ -> default in
+  let safe_filter f = List.filter (default_on_exn f false) in
+  let safe_map f = Stdext.Listext.List.filter_map (default_on_exn (fun x -> Some (f x)) None) in
+  let safe_flat_map f l = safe_map (default_on_exn f []) l |> List.flatten in
+
   let sr = Db.VDI.get_SR ~__context ~self in
   let hosts_with_attached_pbds =
     Db.SR.get_PBDs ~__context ~self:sr
-    |> List.filter (fun pbd -> Db.PBD.get_currently_attached ~__context ~self:pbd)
-    |> List.map (fun pbd -> Db.PBD.get_host ~__context ~self:pbd)
+    |> safe_filter (fun pbd -> Db.PBD.get_currently_attached ~__context ~self:pbd)
+    |> safe_map (fun pbd -> Db.PBD.get_host ~__context ~self:pbd)
```

Comment:
I was considering using `Db.PBD.get_refs_where` here instead, to make it shorter. That way I could also get rid of the `safe_filter` function.